### PR TITLE
Fix inconsistent subscription generator clean up when the client disconnects

### PR DESCRIPTION
**Pull request by wuyuanyi135**
## Description

1. catch the `asyncio.CancelledError` raised when the client disconnects. Otherwise, the task will exit and leave un-retrieved exception error.
2. The checks on L202 seems unable to catch the disconnected state and insist to send the completion to the disconnected socket, leading to an error printed in the console. I wonder whether we could completely remove the `if` check and just use the try-except?
3. Unit-test pending

## Types of Changes

- [ ] Core
- [x] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Documentation

## Issues Fixed or Closed by This PR
#889 

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: strawberry/asgi/__init__.py
Comment (BryceBeagle) [inline strawberry/asgi/__init__.py:200]: Can you make sure you've [installed the `pre-commit` hooks](https://github.com/strawberry-graphql/strawberry#pre-commit)? There's some minor formatting issues here, and the imports haven't been sorted.

**Review by BryceBeagle — CHANGES_REQUESTED** [review]

**Comment by botberry** [comment]
Hi, thanks for contributing to Strawberry 🍓!

We noticed that this PR is missing a `RELEASE.md` file. We use that to automatically do releases here on GitHub and, most importantly, to PyPI!

So as soon as this PR is merged, a release will be made 🚀.

Here's an example of `RELEASE.md`:

```markdown
Release type: patch

Description of the changes, ideally with some examples, if adding a new feature.
```

Release type can be one of patch, minor or major. We use [semver](https://semver.org/),so make sure to pick the appropriate type. If in doubt feel free to ask :)

File: strawberry/asgi/__init__.py
Comment (wuyuanyi135) [reply strawberry/asgi/__init__.py:200]: Sorry, I probably forgot that. I will first settle the PR on `graphql-core` then get back here.

**Comment by wuyuanyi135** [comment]
We will wait for the upstream to decide whether the CancelledError should interrupt the MapAsyncIterator with StopAsyncIteration (normal exit) or some other exception. If this will be deemed as a normal exit we will consider put a return after the `async for`.

File: strawberry/asgi/__init__.py
Comment (patrick91) [reply strawberry/asgi/__init__.py:200]: @wuyuanyi135 does that mean we need to wait for the updated release of `graphql-core` to merge this PR?

Comment (wuyuanyi135) [reply strawberry/asgi/__init__.py:200]: @patrick91 The fix has been merged in `graphql-core`. We do need the updated release here.

File: strawberry/asgi/__init__.py
Comment (marcoacierno) [inline strawberry/asgi/__init__.py:199]:
```suggestion
            # Suppress the error and prevent continuing to the sending completion message stage
```
?

File: strawberry/asgi/__init__.py
Comment (marcoacierno) [inline strawberry/asgi/__init__.py:209]: I think we should catch ConnectionClosedError on L198 - L196 as well? 
There is a `except Exception as error` which might capture this error as well and try to send it as error to the client

File: strawberry/asgi/__init__.py
Comment (marcoacierno) [inline strawberry/asgi/__init__.py:197]: I think this should come before `except Exception as error:`?

here https://docs.python.org/3.8/library/asyncio-exceptions.html it says that it's using BaseException on 3.8 as base so it will be captured by `except Exception` above?

File: strawberry/asgi/__init__.py
Comment (DoctorJohn) [reply strawberry/asgi/__init__.py]: @wuyuanyi135 Strawberry was upgraded to graphql-core 3.1.5 (24080d57817dabd1e571f5463a82e4eeaf0b2dde) which includes your upstream changes.

File: strawberry/asgi/__init__.py
Comment (DoctorJohn) [reply strawberry/asgi/__init__.py:214]: Just out of curiosity: where could `websockets.ConnectionClosedError` be raised?

I noticed `self._send_message` calls `starlette.WebSocket.send_json` calls `starlette.WebSocket.send`, but these methods appear to only raise `RuntimeError`s. (Finally `starlette.WebSocket.send` calls the `send` method passed to `strawberry.asgi.GraphQL.__call__`, thats where I lost track).

**Comment by wuyuanyi135** [comment]
@patrick91 Sorry for the late response. I wonder what needs to be worked on in this PR now?

**Comment by patrick91** [comment]
> @patrick91 Sorry for the late response. I wonder what needs to be worked on in this PR now?

I wanted to add some tests, but didn’t manage to do so, are you able to write one? 😌

**Comment by wuyuanyi135** [comment]
I will work on this one tomorrow.

**Pull request closed without merging**
